<a href="https://colab.research.google.com/github/HO94/InsightMining_HonGong/blob/main/07_2_%EC%8B%AC%EC%B8%B5_%EC%8B%A0%EA%B2%BD%EB%A7%9D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

입력층   = 일차원 데이터 셋(784개 뉴런)  
2개의 층 = 은닉층, 출력층

이진 분류 : 시그모이드  함수  
다중 분류 : 소프트맥스 함수

은닉층 + 활성화 함수 -> 비선형 함수로 변형, 단순한 선형으로 나타나지 않도록 함.     
활성화 함수가 없다면 출력층과 합쳐서 하나의 층으로 나타남

In [ ]:
from tensorflow import keras
(train_input, train_target), (test_input, test_target) = keras.datasets.fashion_mnist.load_data()

4423680/4422102 [==============================] - 0s 0us/step


In [ ]:
from sklearn.model_selection import train_test_split
train_scaled = train_input / 255.0
train_scaled = train_scaled.reshape(-1, 28*28)
train_scaled, val_scaled, train_target, val_target = train_test_split(train_scaled, train_target, test_size = 0.2, random_state = 42)

## 심층 신경망

In [ ]:
dense1 = keras.layers.Dense(100, activation = "sigmoid", input_shape = (784,)) # 은닉층
dense2 = keras.layers.Dense(10, activation = "softmax") # 출력층

model = keras.Sequential([dense1, dense2])

## summary() 메서드

In [ ]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 100)               78500     
_________________________________________________________________
dense_1 (Dense)              (None, 10)                1010      
Total params: 79,510
Trainable params: 79,510
Non-trainable params: 0
_________________________________________________________________


- 입력층 784개 뉴런 * 은닉층 100개 뉴런   (완전연결, 가중치) + 100개 뉴런 (절편)
= 78500개  

- 은닉층 100개 뉴련 * 출력층 10개 뉴런 + 10개 뉴런 = 1010개  
  
  

fit(...batchsize = 32) -> 미니배치로 적용  
32개씩 잘라서 모델 훈련  
모델을 만들 때에는 변경 x  
훈련 시킬 때 자연스럽게 변경

## 층을 추가하는 다른 방법

In [ ]:
# 1번
model = keras.Sequential([keras.layers.Dense(100, activation = "sigmoid", input_shape = (784,), name = "hidden"),
                          keras.layers.Dense(10, activation = "softmax", name = "output")],
                          name = "패션 MNIST 모델")

In [ ]:
# 2번 (실제로 많이 사용하는 코드, add 메서드 사이 if문 사용 등 유연한 모델 사용 가능)
model = keras.Sequential()
model.add(keras.layers.Dense(100, activation = "sigmoid", input_shape = (784,)))
model.add(keras.layers.Dense(10, activation = "softmax"))

## 모델 훈련

In [ ]:
model.compile(loss = "sparse_categorical_crossentropy", metrics = "accuracy")

model.fit(train_scaled, train_target, epochs = 5)

Epoch 1/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.7567 - accuracy: 0.7591
Epoch 2/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.4148 - accuracy: 0.8509
Epoch 3/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.3771 - accuracy: 0.8637
Epoch 4/5
1500/1500 [==============================] - 4s 2ms/step - loss: 0.3581 - accuracy: 0.8712
Epoch 5/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.3397 - accuracy: 0.8778


## 렐루 함수와 Flatten 층

시그모이드 함수의 단점  
- 선형 출력값이 극단값에 갈수록 값의 변화가 매우 작아짐(포화), 신경망 모델이 빠르게 학습할 수 없어 층을 깊게 쌓지 못함  

렐루 함수 사용  
- max 함수  
- 뉴런의 출력함수(z)가 0보다 크면 z출력, 0보다 작으면 버리는 함수

새로운 층 Flatten  
- 실제 신경망 모델, 알고리즘에 등장x
- 편의로 만든 층  
- 케라스에만 존재하는 유틸리티 층  
- keras 모듈 밑에 있기 때문에 층이라 부름  
- params = 0  
- 28*28 =784(일차원 배열)로 펼치는 작업을 대신 해줌  
- 필요한 전처리 과정을 모델 안에 포함시키기 위해


In [ ]:
model = keras.Sequential()

model.add(keras.layers.Flatten(input_shape = (28,28)))
model.add(keras.layers.Dense(100, activation = "relu"))
model.add(keras.layers.Dense(10, activation = "softmax"))

In [ ]:
model.summary()

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten (Flatten)            (None, 784)               0         
_________________________________________________________________
dense_5 (Dense)              (None, 100)               78500     
_________________________________________________________________
dense_6 (Dense)              (None, 10)                1010      
Total params: 79,510
Trainable params: 79,510
Non-trainable params: 0
_________________________________________________________________


In [ ]:
(train_input, train_target), (test_input, test_target) = keras.datasets.fashion_mnist.load_data()

from sklearn.model_selection import train_test_split
train_scaled = train_input / 255.0
train_scaled, val_scaled, train_target, val_target = train_test_split(train_scaled, train_target, test_size = 0.2, random_state = 42)

In [ ]:
model.compile(loss = "sparse_categorical_crossentropy", metrics = "accuracy")

model.fit(train_scaled, train_target, epochs = 5)

Epoch 1/5
1500/1500 [==============================] - 4s 2ms/step - loss: 0.6591 - accuracy: 0.7701
Epoch 2/5
1500/1500 [==============================] - 4s 2ms/step - loss: 0.3991 - accuracy: 0.8577
Epoch 3/5
1500/1500 [==============================] - 4s 2ms/step - loss: 0.3518 - accuracy: 0.8746
Epoch 4/5
1500/1500 [==============================] - 4s 2ms/step - loss: 0.3277 - accuracy: 0.8820
Epoch 5/5
1500/1500 [==============================] - 4s 2ms/step - loss: 0.3168 - accuracy: 0.8849


In [ ]:
model.evaluate(val_scaled, val_target)

375/375 [==============================] - 1s 1ms/step - loss: 0.3789 - accuracy: 0.8724


[0.3788610100746155, 0.8724166750907898]

## 옵티마이저

compile 메서드 optimizer = "sgd" -> 확률적 경사하강법  

optimizer = "sgd" : 미니배치로 사용됨  

최적화 옵티마이저 종류가 다양함  
기본 경사 하강법 옵티마이저  
- SGD -> 모멘텀 -> 네스테로프 모멘텀  

적응적 학습률 옵티마이저  
- RMSprop -> Adam
- 학습률이 점차 변화

In [ ]:
#동일한 방법
model.compile(optimizer = "sgd", loss = "sparse_categorical_crossentropy", metrics = "accuracy")

sgd = keras.optimizers.SGD()
model.compile(optimizer = sgd, loss = "sparse_categorical_crossentropy", metrics = "accuracy")

In [ ]:
# 매개변수 값을 변화하고 싶다면
sgd = keras.optimizers.SGD(learning_rate = 0.1) # learning_rate 기본값 0.01
sgd = keras.optimizers.SGD(momentum = 0.9, nesterov = True) # 모멘텀 방법 사용(기본값 0), 네스테로프 사용(기본값 False)

In [ ]:
# 적응적 학습률 - Adagrad
adagrad = keras.optimizers.Adagrad()
model.compile(optimizer = adagrad, loss = "sparse_categorical_crossentropy", metrics = "accuracy")

In [ ]:
# 적응적 학습률 - RMSprop
rmsprop = keras.optimizers.RMSprop()
model.compile(optimizer = rmsprop, loss = "sparse_categorical_crossentropy", metrics = "accuracy")

model.fit(train_scaled, train_target, epochs = 5)

Epoch 1/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.3136 - accuracy: 0.8874
Epoch 2/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.2918 - accuracy: 0.8954
Epoch 3/5
1500/1500 [==============================] - 4s 3ms/step - loss: 0.2788 - accuracy: 0.8986
Epoch 4/5
1500/1500 [==============================] - 4s 2ms/step - loss: 0.2858 - accuracy: 0.8994
Epoch 5/5
1500/1500 [==============================] - 4s 2ms/step - loss: 0.2652 - accuracy: 0.9043


### 적응적 학습률 옵티마이저

In [ ]:
# 입문서라 수식 제외
model = keras.Sequential()
model.add(keras.layers.Flatten(input_shape = (28, 28)))
model.add(keras.layers.Dense(100, activation = "relu"))
model.add(keras.layers.Dense(10, activation = "softmax"))


In [ ]:
model.compile(optimizer = "adam", loss = "sparse_categorical_crossentropy", metrics = "accuracy")
model.fit(train_scaled, train_target, epochs = 5)

model.evaluate(val_scaled, val_target)

Epoch 1/5
1500/1500 [==============================] - 4s 2ms/step - loss: 0.6866 - accuracy: 0.7649
Epoch 2/5
1500/1500 [==============================] - 3s 2ms/step - loss: 0.4069 - accuracy: 0.8553
Epoch 3/5
1500/1500 [==============================] - 3s 2ms/step - loss: 0.3620 - accuracy: 0.8708
Epoch 4/5
1500/1500 [==============================] - 3s 2ms/step - loss: 0.3286 - accuracy: 0.8807
Epoch 5/5
375/375 [==============================] - 1s 1ms/step - loss: 0.3316 - accuracy: 0.8834


[0.3316112458705902, 0.8834166526794434]

신경망 층이 특성공학 역할을 많은 부분 해주기 때문에 이미지, 텍스트 등에서 특성을 찾을 수 있음(자동으로 특성을 학습)  

특성 공학 : 특성을 수동으로 뽑아야 함